File: client/lib/categories.ts
export const SYSTEM_NEWS_CATEGORIES = [
  "amni",
  "ciyaaro",
  "siyaasad",
  "caalamka",
] as const;

export type SystemNewsCategory = (typeof SYSTEM_NEWS_CATEGORIES)[number];

const DEFAULT_NEWS_CATEGORY: SystemNewsCategory = "caalamka";

const CATEGORY_LABELS: Record<SystemNewsCategory, string> = {
  amni: "Amni",
  ciyaaro: "Ciyaaro",
  siyaasad: "Siyaasad",
  caalamka: "Caalamka",
};

const CATEGORY_ALIASES: Record<string, SystemNewsCategory> = {
  amni: "amni",
  security: "amni",
  safety: "amni",
  defense: "amni",
  defence: "amni",
  ciyaaro: "ciyaaro",
  sports: "ciyaaro",
  sport: "ciyaaro",
  siyaasad: "siyaasad",
  politics: "siyaasad",
  political: "siyaasad",
  caalamka: "caalamka",
  world: "caalamka",
  international: "caalamka",
  global: "caalamka",
  unknown: "caalamka",
  business: "caalamka",
  other: "caalamka",
  general: "caalamka",
};

export function normalizeCategory(category: string | null | undefined) {
  const cleaned = (category ?? "").trim().toLowerCase();
  if ((SYSTEM_NEWS_CATEGORIES as readonly string[]).includes(cleaned)) {
    return cleaned as SystemNewsCategory;
  }

  if (cleaned in CATEGORY_ALIASES) {
    return CATEGORY_ALIASES[cleaned];
  }

  for (const valid of SYSTEM_NEWS_CATEGORIES) {
    if (cleaned.includes(valid)) {
      return valid;
    }
  }

  return DEFAULT_NEWS_CATEGORY;
}

export function getCategoryLabel(category: string) {
  return CATEGORY_LABELS[normalizeCategory(category)];
}

export function getNewsCategories() {
  return [...SYSTEM_NEWS_CATEGORIES];
}

File: services/categories.py
SYSTEM_NEWS_CATEGORIES = ("amni", "ciyaaro", "siyaasad", "caalamka")
DEFAULT_NEWS_CATEGORY = "caalamka"

CATEGORY_ALIASES = {
    "amni": "amni",
    "security": "amni",
    "safety": "amni",
    "defense": "amni",
    "defence": "amni",
    "ciyaaro": "ciyaaro",
    "sports": "ciyaaro",
    "sport": "ciyaaro",
    "siyaasad": "siyaasad",
    "politics": "siyaasad",
    "political": "siyaasad",
    "caalamka": "caalamka",
    "world": "caalamka",
    "international": "caalamka",
    "global": "caalamka",
    "unknown": "caalamka",
    "business": "caalamka",
    "other": "caalamka",
    "general": "caalamka",
}


def get_system_news_categories() -> list[str]:
    return list(SYSTEM_NEWS_CATEGORIES)


def normalize_category(category: str | None) -> str:
    if not category or not str(category).strip():
        return DEFAULT_NEWS_CATEGORY

    cleaned = str(category).strip().lower()
    if cleaned in SYSTEM_NEWS_CATEGORIES:
        return cleaned

    mapped = CATEGORY_ALIASES.get(cleaned)
    if mapped:
        return mapped

    for valid in SYSTEM_NEWS_CATEGORIES:
        if valid in cleaned:
            return valid

    return DEFAULT_NEWS_CATEGORY


def get_category_query_values(category: str | None) -> list[str] | None:
    if not category or not str(category).strip():
        return None

    cleaned = str(category).strip().lower()
    canonical = None
    if cleaned in SYSTEM_NEWS_CATEGORIES:
        canonical = cleaned
    elif cleaned in CATEGORY_ALIASES:
        canonical = CATEGORY_ALIASES[cleaned]

    if canonical is None:
        return [cleaned]

    values = {canonical}
    for alias, mapped in CATEGORY_ALIASES.items():
        if mapped == canonical:
            values.add(alias)
    return list(values)
